### PR TITLE
DOCSUP-13429: Document the short_circuit_function_evaluation setting

### PR DESCRIPTION
**Pull request by sevirov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:

- Documentation for #23367

### CODE REVIEW

File: docs/en/operations/settings/settings.md [hunk 1]
```diff
@@ -3475,3 +3475,17 @@ Possible values:
...
 
 Default value: `1000`.
+
+## short_circuit_function_evaluation {#short-circuit-function-evaluation}
+
+Allows calculating the [if](../../sql-reference/functions/conditional-functions.md#if), [multiIf](../../sql-reference/functions/conditional-functions.md#multiif), [and](../../sql-reference/functions/logical-functions.md#logical-and-function) and [or](../../sql-reference/functions/logical-functions.md#logical-or-function) functions according to a [short scheme](https://en.wikipedia.org/wiki/Short-circuit_evaluation).
+
+Possible values:
+
+- `enable ` — Enables short-circuit function evaluation for functions that are suitable for it (can throw or computationally heavy).
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3485]:
```suggestion
- `enable` — Enables short-circuit function evaluation for functions that are suitable for it (can throw or computationally heavy).
```

File: docs/en/operations/settings/settings.md [hunk 2]
```diff
@@ -3475,3 +3475,17 @@ Possible values:
...
+- `force_enable` — Enables short-circuit function evaluation for all functions.
+- `disable` — Disables short-circuit function evaluation.
+
+Default value: `enable`.
+
+This is necessary to optimize the execution of complex expressions in these functions and to prevent possible exceptions (such as division by zero, when it is not expected).
```
Comment (olgarev) [inline docs/en/operations/settings/settings.md:3491]:
```suggestion
This helps optimize the execution of complex expressions in these functions and prevent possible exceptions (such as division by zero when it is not expected).
```

File: docs/en/sql-reference/functions/logical-functions.md [hunk 3]
```diff
@@ -19,6 +19,8 @@ Calculates the result of the logical conjunction between two or more values. Cor
 and(val1, val2...)
 ```
 
+You can use the [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation) setting to calculate the `and` function according to a short scheme. So `vali` will be evaluated only on rows where `(val1 AND val2 AND ... AND val{i-1})` is true. For example, an exception about division by zero is not thrown when executing the query `SELECT number != 0 AND intDiv(1, number) == 0 AND number != 2 AND intDiv(1, number - 2) == 0 FROM numbers(10)`.
```
Comment (olgarev) [inline docs/en/sql-reference/functions/logical-functions.md:22]:
```suggestion
You can use the [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation) setting to calculate the `and` function according to a short scheme. If this setting is enabled, `vali` is evaluated only on rows where `(val1 AND val2 AND ... AND val{i-1})` is true. For example, an exception about division by zero is not thrown when executing the query `SELECT number != 0 AND intDiv(1, number) = 0 AND number != 2 AND intDiv(1, number - 2) = 0 FROM numbers(10)`.
```

File: docs/ru/operations/settings/settings.md [hunk 4]
```diff
@@ -3376,6 +3376,18 @@ SETTINGS index_granularity = 8192 │
 
 Значение по умолчанию: `1000`.
 
+## short_circuit_function_evaluation {#short-circuit-function-evaluation}
+
+Позволяет вычислять функции [if](../../sql-reference/functions/conditional-functions.md#if), [multiIf](../../sql-reference/functions/conditional-functions.md#multiif), [and](../../sql-reference/functions/logical-functions.md#logical-and-function) и [or](../../sql-reference/functions/logical-functions.md#logical-or-function) по [короткой схеме](https://ru-wikipedia-org.turbopages.org/ru.wikipedia.org/s/wiki/Вычисления_по_короткой_схеме). Это помогает оптимизировать выполнение сложных выражений в этих функциях и предотвратить возможные исключения (как, например, деление на ноль, когда оно не ожидается).
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:3381]:
```suggestion
Позволяет вычислять функции [if](../../sql-reference/functions/conditional-functions.md#if), [multiIf](../../sql-reference/functions/conditional-functions.md#multiif), [and](../../sql-reference/functions/logical-functions.md#logical-and-function) и [or](../../sql-reference/functions/logical-functions.md#logical-or-function) по [короткой схеме](https://ru-wikipedia-org.turbopages.org/ru.wikipedia.org/s/wiki/Вычисления_по_короткой_схеме). Это помогает оптимизировать выполнение сложных выражений в этих функциях и предотвратить возможные исключения (например, деление на ноль, когда оно не ожидается).
```
```suggestion
Позволяет вычислять функции [if](../../sql-reference/functions/conditional-functions.md#if), [multiIf](../../sql-reference/functions/conditional-functions.md#multiif), [and](../../sql-reference/functions/logical-functions.md#logical-and-function) и [or](../../sql-reference/functions/logical-functions.md#logical-or-function) по [короткой схеме](https://ru-wikipedia-org.turbopages.org/ru.wikipedia.org/s/wiki/Вычисления_по_короткой_схеме). Это помогает оптимизировать выполнение сложных выражений в этих функциях и предотвратить возможные исключения (как, например, деление на ноль, когда оно не ожидается).
```

Comment (olgarev) [inline docs/ru/operations/settings/settings.md:3381]:
```suggestion
Позволяет вычислять функции [if](../../sql-reference/functions/conditional-functions.md#if), [multiIf](../../sql-reference/functions/conditional-functions.md#multiif), [and](../../sql-reference/functions/logical-functions.md#logical-and-function) и [or](../../sql-reference/functions/logical-functions.md#logical-or-function) по [короткой схеме](https://ru-wikipedia-org.turbopages.org/ru.wikipedia.org/s/wiki/Вычисления_по_короткой_схеме). Это помогает оптимизировать выполнение сложных выражений в этих функциях и предотвратить возможные исключения (например, деление на ноль, когда оно не ожидается).
```
```suggestion
Позволяет вычислять функции [if](../../sql-reference/functions/conditional-functions.md#if), [multiIf](../../sql-reference/functions/conditional-functions.md#multiif), [and](../../sql-reference/functions/logical-functions.md#logical-and-function) и [or](../../sql-reference/functions/logical-functions.md#logical-or-function) по [короткой схеме](https://ru-wikipedia-org.turbopages.org/ru.wikipedia.org/s/wiki/Вычисления_по_короткой_схеме). Это помогает оптимизировать выполнение сложных выражений в этих функциях и предотвратить возможные исключения (как, например, деление на ноль, когда оно не ожидается).
```

File: docs/ru/operations/settings/settings.md [hunk 5]
```diff
@@ -3376,6 +3376,18 @@ SETTINGS index_granularity = 8192 │
...
+
+Возможные значения:
+
+-   `enable` — функции, которые подходят для этого, вычисляются по короткой схеме (может сгенерировать исключение или быть сложным с вычислительной точки зрения).
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:3385]:
```suggestion
-   `enable` — по короткой схеме вычисляются функции, которые подходят для этого (могут сгенерировать исключение или требуют сложных вычислений).
```
```suggestion
-   `enable` — функции, которые подходят для этого, вычисляются по короткой схеме (может сгенерировать исключение или быть сложным с вычислительной точки зрения).
```

File: docs/ru/operations/settings/settings.md [hunk 6]
```diff
@@ -3376,6 +3376,18 @@ SETTINGS index_granularity = 8192 │
...
+-   `force_enable` — все функции вычисляются по короткой схеме.
+-   `disable` — отключено вычисление функций по короткой схеме.
```
Comment (olgarev) [inline docs/ru/operations/settings/settings.md:3387]:
```suggestion
-   `disable` — вычисление функций по короткой схеме отключено.
```
```suggestion
-   `disable` — отключено вычисление функций по короткой схеме.
```

File: docs/ru/sql-reference/functions/conditional-functions.md [hunk 7]
```diff
@@ -12,11 +12,13 @@ toc_title: "Условные функции"
 **Синтаксис**
 
 ``` sql
-SELECT if(cond, then, else)
+if(cond, then, else)
 ```
 
 Если условие `cond` не равно нулю, то возвращается результат выражения `then`. Если условие `cond` равно нулю или является NULL, то результат выражения `then` пропускается и возвращается результат выражения `else`.
 
+Чтобы вычислять функцию `if` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `then` вычисляется только для строк, где условие `cond` верно, а выражение `else` – где условие `cond` неверно. Например, при выполнении запроса `SELECT if(number = 0, 0, intDiv(42, number)) FROM numbers(10)` не будет сгенерировано исключение о делении на ноль, так как `intDiv(42, number)` будет вычислено только для чисел, которые не удовлетворяют условию `number = 0`.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/conditional-functions.md:20]:
```suggestion
Чтобы вычислять функцию `if` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `then` вычисляется только для строк, где условие `cond` верно, а выражение `else` – для строк, где условие `cond` неверно. Например, при выполнении запроса `SELECT if(number = 0, 0, intDiv(42, number)) FROM numbers(10)` не будет сгенерировано исключение из-за деления на ноль, так как `intDiv(42, number)` будет вычислено только для чисел, которые не удовлетворяют условию `number = 0`.
```
```suggestion
Чтобы вычислять функцию `if` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `then` вычисляется только для строк, где условие `cond` верно, а выражение `else` – где условие `cond` неверно. Например, при выполнении запроса `SELECT if(number = 0, 0, intDiv(42, number)) FROM numbers(10)` не будет сгенерировано исключение о делении на ноль, так как `intDiv(42, number)` будет вычислено только для чисел, которые не удовлетворяют условию `number = 0`.
```

Comment (olgarev) [inline docs/ru/sql-reference/functions/conditional-functions.md:20]:
```suggestion
Чтобы вычислять функцию `if` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `then` вычисляется только для строк, где условие `cond` верно, а выражение `else` – для строк, где условие `cond` неверно. Например, при выполнении запроса `SELECT if(number = 0, 0, intDiv(42, number)) FROM numbers(10)` не будет сгенерировано исключение из-за деления на ноль, так как `intDiv(42, number)` будет вычислено только для чисел, которые не удовлетворяют условию `number = 0`.
```

File: docs/ru/sql-reference/functions/conditional-functions.md [hunk 8]
```diff
@@ -77,7 +79,13 @@ SELECT if(0, plus(2, 2), plus(2, 6));
...
 Позволяет более компактно записать оператор [CASE](../operators/index.md#operator_case) в запросе.
 
-    multiIf(cond_1, then_1, cond_2, then_2...else)
+**Синтаксис**
+
+``` sql
+multiIf(cond_1, then_1, cond_2, then_2, ..., else)
+```
+
+Чтобы вычислять функцию `multiIf` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `then_i` вычисляется только для строк, где условие `((NOT cond_1) AND (NOT cond_2) AND ... AND (NOT cond_{i-1}) AND cond_i)` верно, `cond_i` вычисляется только для строк, где условие `((NOT cond_1) AND (NOT cond_2) AND ... AND (NOT cond_{i-1}))` верно. Например, при выполнении запроса `SELECT multiIf(number = 0, 0, number = 1, intDiv(1, number), number = 2, intDiv(1, number - 1), number = 3, intDiv(1, number - 2), intDiv(1, number - 3)) FROM numbers(10)` не будет сгенерировано исключение о делении на ноль.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/conditional-functions.md:88]:
```suggestion
Чтобы вычислять функцию `multiIf` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `then_i` вычисляется только для строк, где условие `((NOT cond_1) AND (NOT cond_2) AND ... AND (NOT cond_{i-1}) AND cond_i)` верно, `cond_i` вычисляется только для строк, где условие `((NOT cond_1) AND (NOT cond_2) AND ... AND (NOT cond_{i-1}))` верно. Например, при выполнении запроса `SELECT multiIf(number = 0, 0, number = 1, intDiv(1, number), number = 2, intDiv(1, number - 1), number = 3, intDiv(1, number - 2), intDiv(1, number - 3)) FROM numbers(10)` не будет сгенерировано исключение из-за деления на ноль.
```

File: docs/ru/sql-reference/functions/logical-functions.md [hunk 9]
```diff
@@ -19,6 +19,8 @@ toc_title: "Логические функции"
 and(val1, val2...)
 ```
 
+Чтобы вычислять функцию `and` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `vali` вычисляется только для строк, где условие `(val1 AND val2 AND ... AND val{i-1})` верно. Например, при выполнении запроса `SELECT number != 0 AND intDiv(1, number) = 0 AND number != 2 AND intDiv(1, number - 2) = 0 FROM numbers(10)` не будет сгенерировано исключение о делении на ноль.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/logical-functions.md:22]:
```suggestion
Чтобы вычислять функцию `and` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `vali` вычисляется только для строк, где условие `(val1 AND val2 AND ... AND val{i-1})` верно. Например, при выполнении запроса `SELECT number != 0 AND intDiv(1, number) = 0 AND number != 2 AND intDiv(1, number - 2) = 0 FROM numbers(10)` не будет сгенерировано исключение из-за деления на ноль.
```
```suggestion
Чтобы вычислять функцию `and` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `vali` вычисляется только для строк, где условие `(val1 AND val2 AND ... AND val{i-1})` верно. Например, при выполнении запроса `SELECT number != 0 AND intDiv(1, number) = 0 AND number != 2 AND intDiv(1, number - 2) = 0 FROM numbers(10)` не будет сгенерировано исключение о делении на ноль.
```

File: docs/ru/sql-reference/functions/logical-functions.md [hunk 10]
```diff
@@ -71,6 +73,8 @@ SELECT and(NULL, 1, 10, -2);
 and(val1, val2...)
 ```
 
+Чтобы вычислять функцию `or` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `vali` вычисляется только для строк, где условие `((NOT val1) AND (NOT val2) AND ... AND (NOT val{i-1}))` верно. Например, при выполнении запроса `SELECT number = 0 OR intDiv(1, number) != 0 OR number = 2 OR intDiv(1, number - 2) != 0 FROM numbers(10)` не будет сгенерировано исключение о делении на ноль.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/logical-functions.md:76]:
```suggestion
Чтобы вычислять функцию `or` по короткой схеме, используйте настройку [short_circuit_function_evaluation](../../operations/settings/settings.md#short-circuit-function-evaluation). Если настройка включена, то выражение `vali` вычисляется только для строк, где условие `((NOT val1) AND (NOT val2) AND ... AND (NOT val{i-1}))` верно. Например, при выполнении запроса `SELECT number = 0 OR intDiv(1, number) != 0 OR number = 2 OR intDiv(1, number - 2) != 0 FROM numbers(10)` не будет сгенерировано исключение из-за деления на ноль.
```

**Comment by sevirov** [comment]
@Avogar Please, review this PR.

**Review by Avogar — APPROVED** [review]

**Pull request merged**